File: src/memory.c
/*
 * Copyright (C) 2013 Omar Sandoval
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <assert.h>
#include <ctype.h>
#include <errno.h>
#include <inttypes.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <sys/ptrace.h>

#include "memory.h"

static void dump_decimal(unsigned char *buf, size_t size);
static void dump_unsigned_decimal(unsigned char *buf, size_t size);
static void dump_octal(unsigned char *buf, size_t size);
static void dump_hexadecimal(unsigned char *buf, size_t size);
static void dump_float(unsigned char *buf, size_t size);
static void dump_character(unsigned char *buf, size_t size);
static void dump_address(unsigned char *buf, size_t size);
static void escape_character(unsigned char c, char output[5]);

int dump_memory(pid_t pid, void *addr, size_t repeat,
                enum mem_format format, size_t size)
{
    long buffer;

    for (size_t offset = 0; offset < (repeat * size); offset += size) {
        unsigned char *p;

        /* Split up rows */
        size_t row;
        if (format == FMT_OCTAL)
            row = (size <= SZ_HALFWORD) ? 0x8 : 0x10;
        else if (format == FMT_FLOAT)
            row = (size == SZ_WORD) ? 0x10 : 0x20;
        else if (format == FMT_CHARACTER)
            row = 0x8;
        else
            row = (size == SZ_BYTE) ? 0x8 : 0x10;

        if ((offset % row) == 0) {
            if (offset > 0)
                printf("\n");
            printf("%p: ", addr + offset);
        }

        if ((offset % sizeof(long)) == 0) {
            errno = 0;
            buffer = ptrace(PTRACE_PEEKDATA, pid, addr + offset, NULL);
            if (errno) {
                printf("\n");
                fprintf(stderr, "Cannot access memory at address %p\n",
                        addr + offset);
                return 1;
            }
        }

        p = ((unsigned char *) &buffer) + (offset % sizeof(long));

        switch (format) {
            case FMT_DECIMAL:
                dump_decimal(p, size);
                break;
            case FMT_UNSIGNED_DECIMAL:
                dump_unsigned_decimal(p, size);
                break;
            case FMT_OCTAL:
                dump_octal(p, size);
                break;
            case FMT_HEXADECIMAL:
                dump_hexadecimal(p, size);
                break;
            case FMT_FLOAT:
                dump_float(p, size);
                break;
            case FMT_CHARACTER:
                dump_character(p, size);
                break;
            case FMT_ADDRESS:
                dump_address(p, size);
                break;
            default:
                assert(0 && "Invalid format");
                break;
        }
    }
    printf("\n");

    return 0;
}

int dump_strings(pid_t pid, void *addr, size_t repeat)
{
    char escaped[5];
    size_t strings_printed = 0;
    size_t offset = 0;
    long buffer;

    while (strings_printed < repeat) {
        unsigned char *p;

        /* Print address when printing a new string */
        if (offset == 0)
            printf("%p: \"", addr);

        if ((offset % sizeof(long)) == 0) {
            errno = 0;
            buffer = ptrace(PTRACE_PEEKDATA, pid, addr + offset, NULL);
            if (errno) {
                printf("\n");
                fprintf(stderr, "Cannot access memory at address %p\n",
                        addr + offset);
                return 1;
            }
        }

        p = ((unsigned char *) &buffer) + (offset % sizeof(long));

        if (*p) {
            escape_character(*p, escaped);
            printf("%s", escaped);
            ++offset;
        } else {
            printf("\"\n");
            ++strings_printed;
            addr += offset + 1;
            offset = 0;
        }
    }

    return 0;
}

static void dump_decimal(unsigned char *buf, size_t size)
{
    signed long long decimal;

    /*
     * Field widths represent the maximum length of the string plus spacing (2
     * spaces for half-words, 4 for everything else)
     */
    switch (size) {
        case SZ_BYTE:
            decimal = *((int8_t *) buf);
            printf("%-8lld", decimal);
            break;
        case SZ_HALFWORD:
            decimal = *((int16_t *) buf);
            printf("%-8lld", decimal);
            break;
        case SZ_WORD:
            decimal = *((int32_t *) buf);
            printf("%-15lld", decimal);
            break;
        case SZ_GIANT:
            decimal = *((int64_t *) buf);
            printf("%-24lld", decimal);
            break;
    }
}

static void dump_unsigned_decimal(unsigned char *buf, size_t size)
{
    unsigned long long unsigned_decimal;

    /*
     * Same field widths as signed even though we don't have to worry about the
     * negative sign
     */
    switch (size) {
        case SZ_BYTE:
            unsigned_decimal = *((uint8_t *) buf);
            printf("%-8lld", unsigned_decimal);
            break;
        case SZ_HALFWORD:
            unsigned_decimal = *((uint16_t *) buf);
            printf("%-8llu", unsigned_decimal);
            break;
        case SZ_WORD:
            unsigned_decimal = *((uint32_t *) buf);
            printf("%-15llu", unsigned_decimal);
            break;
        case SZ_GIANT:
            unsigned_decimal = *((uint64_t *) buf);
            printf("%-24llu", unsigned_decimal);
            break;
    }
}

static void dump_octal(unsigned char *buf, size_t size)
{
    unsigned long long octal;

    switch (size) {
        case SZ_BYTE:
            octal = *((uint8_t *) buf);
            printf("0%03llo", octal);
            printf("    ");
            break;
        case SZ_HALFWORD:
            octal = *((uint16_t *) buf);
            printf("0%06llo", octal);
            printf("    ");
            break;
        case SZ_WORD:
            octal = *((uint32_t *) buf);
            printf("0%011llo", octal);
            printf("    ");
            break;
        case SZ_GIANT:
            octal = *((uint64_t *) buf);
            printf("0%022llo", octal);
            printf("      ");
            break;
    }
}

static void dump_hexadecimal(unsigned char *buf, size_t size)
{
    unsigned long long hexadecimal;

    switch (size) {
        case SZ_BYTE:
            hexadecimal = *((uint8_t *) buf);
            printf("0x%02llx", hexadecimal);
            printf("    ");
            break;
        case SZ_HALFWORD:
            hexadecimal = *((uint16_t *) buf);
            printf("0x%04llx", hexadecimal);
            printf("  ");
            break;
        case SZ_WORD:
            hexadecimal = *((uint32_t *) buf);
            printf("0x%08llx", hexadecimal);
            printf("    ");
            break;
        case SZ_GIANT:
            hexadecimal = *((uint64_t *) buf);
            printf("0x%016llx", hexadecimal);
            printf("      ");
            break;
    }
}

static void dump_float(unsigned char *buf, size_t size)
{
    double floating = 0.0;

    assert(size == sizeof(float) || size == sizeof(double));

    switch (size) {
        case SZ_WORD:
            floating = *((float *) buf);
            break;
        case SZ_GIANT:
            floating = *((double *) buf);
            break;
    }

    printf("%-16g", floating);
}

static void dump_character(unsigned char *buf, size_t size)
{
    char escaped[7];
    size_t n;

    assert(size == 1);

    /* Escape the character and put it in quotes */
    escape_character(*buf, escaped + 1);
    n = strlen(escaped + 1) + 1;
    escaped[0] = escaped[n] = '\'';
    escaped[n + 1] = '\0';

    printf("%-8s", escaped);
}

static void dump_address(unsigned char *buf, size_t size)
{
    void *addr;
    assert(size == sizeof(void*));
    addr = *((void **) buf);
    printf("%-24p", addr);
}

static void escape_character(unsigned char c, char output[5])
{
    if (c == '\0') /* Null */
        sprintf(output, "\\0");
    else if (c == '\a') /* Bell */
        sprintf(output, "\\a");
    else if (c == '\b') /* Backspace */
        sprintf(output, "\\b");
    else if (c == '\t') /* Horizontal tab */
        sprintf(output, "\\t");
    else if (c == '\n') /* New line */
        sprintf(output, "\\n");
    else if (c == '\v') /* Vertical tab */
        sprintf(output, "\\v");
    else if (c == '\f') /* Form feed */
        sprintf(output, "\\f");
    else if (c == '\r') /* Carriage return */
        sprintf(output, "\\r");
    else if (isprint(c)) /* Other printable characters */
        sprintf(output, "%c", c);
    else /* Anything else should be an escaped hex sequence */
        sprintf(output, "\\x%02hhx", c);
}

File: asmase-cli/memory.js
/*
 * asmase CLI memory formats.
 *
 * Copyright (C) 2017 Omar Sandoval
 *
 * This file is part of asmase.
 *
 * asmase is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * asmase is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with asmase.  If not, see <http://www.gnu.org/licenses/>.
 */

"use strict";

const bigInt = require('big-integer');

function readUInt16LE(memory, offset) {
  return (memory[offset + 1] << 8) + memory[offset];
}

function readUInt32LE(memory, offset) {
  // * 16777216 instead of << 24 because JavaScript always treats the operands
  // of bitwise operators as signed 32-bit integers.
  return ((memory[offset + 3] * 16777216) + (memory[offset + 2] << 16) +
          (memory[offset + 1] << 8) + memory[offset]);
}

// This returns a bigInt rather than a number.
function readUInt64LE(memory, offset) {
  const hi = readUInt32LE(memory, offset + 4);
  const lo = readUInt32LE(memory, offset);
  return bigInt(hi).shiftLeft(32).add(lo);
}

function readInt8LE(memory, offset) {
  return memory[offset] << 24 >> 24;
}

function readInt16LE(memory, offset) {
  return (memory[offset + 1] << 24 >> 16) | memory[offset];
}

function readInt32LE(memory, offset) {
  return ((memory[offset + 3] << 24) | (memory[offset + 2] << 16) |
          (memory[offset + 1] << 8) | memory[offset]);
}

// Like readUInt32LE, this returns a bigInt.
function readInt64LE(memory, offset) {
  const hi = readInt32LE(memory, offset + 4);
  const lo = readUInt32LE(memory, offset);
  return bigInt(hi).shiftLeft(32).or(lo);
}

function escapeChar(c, escape={}) {
  switch (c) {
    case 0x00:
      return '\\0';
    case 0x07:
      return '\\a';
    case 0x08:
      return '\\b';
    case 0x09:
      return '\\t';
    case 0x0a:
      return '\\n';
    case 0x0b:
      return '\\v';
    case 0x0c:
      return '\\f';
    case 0x0d:
      return '\\r';
    default:
      if (escape.doubleQuote && c === 0x22) {
        return '\\"';
      } else if (escape.singleQuote && c === 0x27) {
        return '\\\'';
      } else if (escape.backslash && c === 0x5c) {
        return '\\\\';
      } else if (c >= 0x20 && c <= 0x7e) {
        return String.fromCharCode(c);
      } else {
        return '\\x' + c.toString(16).padStart(2, '0');
      }
  }
};

const memoryFormats = {
  d: {
    1: {
      columns: 8,
      formatter(memory, offset) {
        const n = readInt8LE(memory, offset);
        return n.toString(10).padStart(8);
      },
    },
    2: {
      columns: 8,
      formatter(memory, offset) {
        const n = readInt16LE(memory, offset);
        return n.toString(10).padStart(8);
      },
    },
    4: {
      columns: 4,
      formatter(memory, offset) {
        const n = readInt32LE(memory, offset);
        return n.toString(10).padStart(14);
      },
    },
    8: {
      columns: 2,
      formatter(memory, offset) {
        const n = readInt64LE(memory, offset);
        return n.toString(10).padStart(24);
      },
    },
  },
  u: {
    1: {
      columns: 8,
      formatter(memory, offset) {
        const n = memory[offset];
        return n.toString(10).padStart(8);
      },
    },
    2: {
      columns: 8,
      formatter(memory, offset) {
        const n = readUInt16LE(memory, offset);
        return n.toString(10).padStart(8);
      },
    },
    4: {
      columns: 4,
      formatter(memory, offset) {
        const n = readUInt32LE(memory, offset);
        return n.toString(10).padStart(14);
      },
    },
    8: {
      columns: 2,
      formatter(memory, offset) {
        const n = readUInt64LE(memory, offset);
        return n.toString(10).padStart(24);
      },
    },
  },
  x: {
    1: {
      columns: 8,
      formatter(memory, offset) {
        const n = memory[offset];
        return '    0x' + n.toString(16).padStart(2, '0');
      },
    },
    2: {
      columns: 8,
      formatter(memory, offset) {
        const n = readUInt16LE(memory, offset);
        return '  0x' + n.toString(16).padStart(4, '0');
      },
    },
    4: {
      columns: 4,
      formatter(memory, offset) {
        const n = readUInt32LE(memory, offset);
        return '    0x' + n.toString(16).padStart(8, '0');
      },
    },
    8: {
      columns: 2,
      formatter(memory, offset) {
        const n = readUInt64LE(memory, offset);
        return '      0x' + n.toString(16).padStart(16, '0');
      },
    },
  },
  o: {
    1: {
      columns: 8,
      formatter(memory, offset) {
        const n = memory[offset];
        return '    0' + n.toString(8).padStart(3, '0');
      },
    },
    2: {
      columns: 4,
      formatter(memory, offset) {
        const n = readUInt16LE(memory, offset);
        return '    0' + n.toString(8).padStart(6, '0');
      },
    },
    4: {
      columns: 4,
      formatter(memory, offset) {
        const n = readUInt32LE(memory, offset);
        return '  0' + n.toString(8).padStart(11, '0');
      },
    },
    8: {
      columns: 2,
      formatter(memory, offset) {
        const n = readUInt64LE(memory, offset);
        return '  0' + n.toString(8).padStart(22, '0');
      },
    },
  },
  t: {
    1: {
      columns: 4,
      formatter(memory, offset) {
        const n = memory[offset];
        return '    ' + n.toString(2).padStart(8, '0');
      },
    },
    2: {
      columns: 2,
      formatter(memory, offset) {
        const n = readUInt16LE(memory, offset);
        return '    ' + n.toString(2).padStart(16, '0');
      },
    },
    4: {
      columns: 1,
      formatter(memory, offset) {
        const n = readUInt32LE(memory, offset);
        return '    ' + n.toString(2).padStart(32, '0');
      },
    },
    8: {
      columns: 1,
      formatter(memory, offset) {
        const n = readUInt64LE(memory, offset);
        return '  ' + n.toString(2).padStart(64, '0');
      },
    },
  },
  c: {
    1: {
      columns: 8,
      formatter(memory, offset) {
        const escape = {singleQuote: true, backslash: true};
        return ('\'' + escapeChar(memory[offset], escape) + '\'').padStart(8);
      },
    },
  },
};

module.exports = {
  readUInt16LE,
  readUInt32LE,
  readUInt64LE,
  readInt8LE,
  readInt16LE,
  readInt32LE,
  readInt64LE,
  escapeChar,
  memoryFormats,
};
